src/compiler/crystal/semantic/call_error.cr

```diff
@@ -642,14 +642,26 @@ class Crystal::Call
   def check_recursive_splat_call(a_def, args)
     if a_def.splat_index
       current_splat_type = args.values.last.type
-      if previous_splat_type = program.splat_expansions[a_def]?
-        if current_splat_type.has_in_type_vars?(previous_splat_type)
-          raise "recursive splat expansion: #{previous_splat_type}, #{current_splat_type}, ..."
-        end
+      previous_splat_types = program.splat_expansions[a_def]?
+      if previous_splat_types
+        previous_splat_types.push(current_splat_type)
+      else
+        previous_splat_types = [current_splat_type]
+        program.splat_expansions[a_def] = previous_splat_types
+      end
```

Comment:
Hm. How about 
```
program.splat_expansions[a_def] ||= []
previous_splat_types = program.splat_expansions[a_def]
previous_splat_types.push(current_splat_type)
```
?

Reply:
It could also work, but it dos like two Hash lookups. In my code there's no hash lookup at all (in this branch)

Reply:
Done! (slightly differently)
Thanks!